File: First-Edition/BookSamples/CH08/src/Cellular/Implementation/Automaton.hs
{-# LANGUAGE DataKinds      #-}
{-# LANGUAGE GADTs          #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE TypeFamilies   #-}
module Cellular.Implementation.Automaton where

import Cellular.Language.Automaton

import GHC.TypeLits


-- -- Implementations

data NeighborsCountImpl
  (state :: IState)
  (counts :: [Nat])
type NeighborsCount s c =
  MkCellCondition (NeighborsCountImpl s c)

data AdjacentsLvlImpl
  (lvl :: Nat)
type AdjacentsLvl lvl = MkNeighborhood (AdjacentsLvlImpl lvl)

data StateImpl
  (name :: Symbol)
  (idx  :: Nat)
type State n i = MkState (StateImpl n i)

data RuleImpl
  (name :: Symbol)
  (code :: Symbol)
  (nh   :: INeighborhood)
  (step :: CustomStep)
type Rule n c nh s = MkRule (RuleImpl n c nh s)

File: First-Edition/BookSamples/rosetta-stone/scala3/rsc2p3_type_level_interfaces/src/main/scala/cellular/language/Automaton.scala
package cellular
package language


// Interfaces

sealed trait IState
type MkState[A] <: IState

sealed trait ICellCondition
type MkCellCondition[A] <: ICellCondition

sealed trait INeighborhood
type MkNeighborhood[A] <: INeighborhood

sealed trait IStateTransition
type MkStateTransition[A] <: IStateTransition

sealed trait IStep
type MkStep[A] <: IStep

sealed trait IRule
type MkRule[A] <: IRule
